### PR TITLE
Add predicates and lemmas for subchain, prefix and suffix on seq.v

### PR DESCRIPTION
**Pull request by ed-hermoreyes**
##### Motivation for this change

In a development with our team, we needed this kind of functions. We discussed about this with some math-comp developers who found a possible contribution interesting.

##### Things done/to do

- [x] added corresponding entries in `CHANGELOG_UNRELEASED.md` (do not edit former entries)
- [x] added corresponding documentation in the headers

##### Automatic note to reviewers

Read [this Checklist](https://github.com/math-comp/math-comp/wiki/Checklist-for-following,-reviewing-and-playing-with-a-PR#checklist-for-reviewing-a-pr) and make sure there is a milestone.

Fixes: #621

### CODE REVIEW

**Comment by chdoc** [comment]
Can you name one or two of the developers you talked to, as they might be good candidates for code reviews?

**Comment by ed-hermoreyes** [comment]
We talked to Cyril Cohen.

File: mathcomp/ssreflect/seq.v
Comment (chdoc) [inline mathcomp/ssreflect/seq.v:3403]:
```suggestion
  reflect (exists s2' : seq T, s2 = s1 ++ s2')  (prefix s1 s2).
```

File: mathcomp/ssreflect/seq.v
Comment (chdoc) [inline mathcomp/ssreflect/seq.v:3473]:
```suggestion
Lemma suffixP s1 s2 :
  reflect (exists s2' : seq T, s2 = s2' ++ s1) (suffix s1 s2).
Proof.
apply: (iffP (prefixP _ _)) => [[s2' rev_s2]|[s2' ->]]; exists (rev s2').
  by rewrite -[s2]revK rev_s2 rev_cat revK.
by rewrite rev_cat.
Qed.
```

**Review by chdoc — CHANGES_REQUESTED** [review]
This is already looking pretty decent, but I think that both the collection of lemmas (statements and names) as well as the proofs should be improved before merging. 

The suggestions I made are to be seen as examples to inspire similar simplifications in other places.

**Comment by mgbedmar** [comment]
We've made a general cleanup following your guidelines and we believe the code could be ready now. Let us know if you have any other suggestions.

File: mathcomp/ssreflect/seq.v
Comment (chdoc) [inline mathcomp/ssreflect/seq.v:3528]: This should follow trivially with subchainP, which can be proved just as simply by not relying on this lemma.

File: mathcomp/ssreflect/seq.v
Comment (chdoc) [inline mathcomp/ssreflect/seq.v:3550]:
```suggestion
elim: s2 s1 => /= [s1|y s2' IHs2 s1].
- apply: (iffP idP) => [/eqP ->|[s] [s'] E]; first by exists [::], [::].
  have: @nilp T [::] by [].
  by rewrite E !cat_nilp => /and3P[/nilP-> /nilP-> /nilP->].
apply: (iffP idP).
- case/orP; first by move/prefixP => -[s ->]; exists [::],s.
  by move/IHs2 => [s [s' ->]]; exists (y::s), s'. 
move => [s [s']]. case: s => //= [|x u]. 
case: s1 => //= ? u [<- ->];by rewrite eqxx prefix_prefix.
by move => [-> E]; rewrite (introT (IHs2 _)) ?orbT // E; exists u, s'.
```

File: mathcomp/ssreflect/seq.v
Comment (chdoc) [inline mathcomp/ssreflect/seq.v:3669]:
```suggestion
Proof. by rewrite -{2}[s](cat_take_drop i) prefix_prefix. Qed.
```

**Review by chdoc — CHANGES_REQUESTED** [review]
This is already looking better, but there is still quite a bit of room for improvement.

Also, please do not merge 'master' into PRs, instead please crate pull requests from branches that you can freely modify (i.e., usually something other than master) and then just rebase onto master, squashing intermediate commits in the process. 

Again, my suggestions are not meant to be exhaustive, but rather as examples to be applied throughout.

File: mathcomp/ssreflect/seq.v
Comment (chdoc) [inline mathcomp/ssreflect/seq.v:3378]:
```suggestion
Proof. by elim: s1 s2 => [[]//|x s IHs s2 /=]; rewrite eqxx IHs.
```

File: mathcomp/ssreflect/seq.v
Comment (chdoc) [inline mathcomp/ssreflect/seq.v:3465]:
```suggestion
Lemma prefix_singleton s x : 
  prefix s [:: x] = (s == [::]) || (s == [:: x]). 
Proof. by case: s => //= y s; rewrite prefixseq0 eqseq_cons. Qed.

Lemma subchain_singleton s x : 
  subchain s [:: x] = (s == [::]) || (s == [:: x]).
Proof. by rewrite /= prefix_singleton orbC orbA orbb. Qed.
```

File: mathcomp/ssreflect/seq.v
Comment (chdoc) [inline mathcomp/ssreflect/seq.v:3661]: I think these can be removed, as they are not needed to prove `take_prefix` and `drop_suffix`

File: mathcomp/ssreflect/seq.v
Comment (chdoc) [inline mathcomp/ssreflect/seq.v:3353]:
```suggestion
Fixpoint subchain_index s1 s2 :=
  if prefix s1 s2 then 0 
  else if s2 is y :: s2' then (subchain_index s1 s2').+1 else 1.
```
I think it is very unfortunate that the result `size s2` can be either the index of the empty subchain or the absence of any valid index. I suggest to change the definition as above, using the unambiguous `(size s2).+1` to signal "no valid index". This changes the lemmas slightly, mostly making them simpler (more expressive). For instance: 
```coq
Lemma subchain_indexseq0 s : subchain_index s [::] = (s != [::]).
Proof. by rewrite /= prefixseq0; case (_ == _). Qed.

Lemma subchain_index_ltn s1 s2 : 
  subchain s1 s2 = (subchain_index s1 s2 < (size s2).+1).
Proof. 
elim: s2 s1 => [|x s2' IH /=] s1; first by rewrite /= prefixseq0; case s1.
by case: (prefix _ _); rewrite IH !ltnS.
Qed.
```

**Review by chdoc — CHANGES_REQUESTED** [review]
I made another pass, and I found even more potential for simplification.

**Comment by mgbedmar** [comment]
> Also, please do not merge 'master' into PRs, instead please crate pull requests from branches that you can freely modify (i.e., usually something other than master) and then just rebase onto master, squashing intermediate commits in the process.

Sorry about that. We're not sure of completely understanding the protocol you just described. Can you confirm that the following steps would fix the current PR?

1. Create another branch `ed-hermoreyes:subchain` with all the commits we've made until now.
2. Re-establish the state of `ed-hermoreyes:master` to be as `math-comp:master`. 
3. Modify this PR so that it goes "into `math-comp:master` from `ed-hermoreyes:subchain`".
4. Rebase squashing all our commits to `ed-hermoreyes:subchain` into a single one.
5. Do 4 each time we request review.

**Comment by chdoc** [comment]
> Sorry about that. We're not sure of completely understanding the protocol you just described. Can you confirm that the following steps would fix the current PR?

No problem. I think this looks like a sound process to fix the current PR, although I do not know how to change the "from" branch for an existing PR in GitHub. 

To be on the safe side, it may be a good idea to implement the changes I proposed before rebase squashing everything. Also, it's not necessary to squash for every review cycle, doing this once when things are (almost) stable can make the the review process more understandable. In the meantime, it may be necessary to rebase (without squashing) in order to resolve conflicts and make the CI run.  In any case, I would suggest only pushing commits you actually want to be reviewed, as every push triggers an email to every member of the core team and a lengthy CI pipeline.

**Comment by chdoc** [comment]
As it comes to the lemma statements and proofs,I'm much happier now than I was initally. And we're down from ~560 to ~330 lines. I will try to make another detailed pass over the weekend. 

I'm still not entirely happy about the naming of things. In particular the double use of `prefix` and `suffix` as indicators for patterns as well as the names of definitions. This leads to confusing names like `prefix_prefix` and `suffix_suffix`. Also, unlike `prefix` and `suffix`, the name `subchain` does not speak to me. Is there maybe an alternative name? 

@CohenCyril Do you maybe have some suggestions?

**Comment by mgbedmar** [comment]
I agree with your objections about the names, we simply didn't find anything better. I even remember that Cyril didn't like `subchain` either. If we keep `prefix` and `suffix` for the names of the predicates, then, following grammar conventions, `subchain` could be renamed as `infix`. But then I have no solution for the double use of `prefix` and `suffix`.

**Comment by chdoc** [comment]
>  If we keep `prefix` and `suffix` for the names of the predicates, then, following grammar conventions, `subchain` could be renamed as `infix`. 

I won't deny that this idea had already crossed my mind. And, after reading about the grammatical term, I'm actually starting to like it better than `subchain`. :thinking:

**Comment by chdoc** [comment]
I made a pass over the proofs, trying to find some synergies by reordering the lemmas. I hope you're fine with me committing directly to the branch of this PR (another reason why it's preferable to use something other than `master` for the source). I think there are more simplifications to be found. 

I removed `subchain_prefixP`, since the RHS of the disjunction subsumes the LHS. 

I still think that some lemmas can be removed, because they are merely compositions (by transitivity) of other lemmas, even though they are not necessarily proved that way currently. Maybe you can make another pass with this in mind.

**Comment by mgbedmar** [comment]
We've been looking at the code. The changes you introduced are nice. Now, to be honest we don't have a clear idea of what the criterion is to decide what lemmas should be kept and which should be removed. The ones that have proofs that are merely compositions of two lemmas are `prefix_subchain` and `suffix_subchain`, but they also exist for `subseq`, so it doesn't feel quite right to remove them. Maybe you can give more concrete directives? Or just remove what you see fit, it's totally okay for you to make changes on the repo.

**Comment by CohenCyril** [comment]
> , although I do not know how to change the "from" branch for an existing PR in GitHub.

The author of the PR might be able to do it using the "edit" button next to the title.

**Comment by chdoc** [comment]
On my own PRs, I can only change the target (base) branch but not the source branch.

**Comment by gares** [comment]
> I have several rephrasings to suggest. I will post in the next few days.

Did you forge to submit the review @CohenCyril ?

**Comment by CohenCyril** [comment]
@CohenCyril I did a big refactoring f852b67 it needs some polishing, double-checking and CHANGELOG update.
@ed-hermoreyes @mgbedmar if you are up for it please proceed, otherwise I will do it on my own time, at some point.

**Comment by chdoc** [comment]
Thanks for getting back to this @CohenCyril . For now I will wait for the authors to comment or for you to make a cleanup pass.

**Comment by CohenCyril** [comment]
@mgbedmar @ed-hermoreyes any update?

**Comment by mgbedmar** [comment]
We have updated the CHANGELOG and made sure everything was on place. For us this is done, let us know if more changes are needed.

**Comment by proux01** [comment]
@chdoc what's the status of this?

**Comment by proux01** [comment]
Will merge next week if no further comments.

**Comment by proux01** [comment]
Sorry, I messed up with this master branch and can no longer push anything on the original [ed-hermoreyes:master](https://github.com/ed-hermoreyes/math-comp) branch, hence reopen the PR. I'm reopening as https://github.com/math-comp/math-comp/pull/902

**Comment by proux01** [comment]
Merged as #902 , sorry again for the mess.

**Pull request closed without merging**
